src/scipp/core/data_group.py

```diff
@@ -0,0 +1,192 @@
+# SPDX-License-Identifier: BSD-3-Clause
+# Copyright (c) 2022 Scipp contributors (https://github.com/scipp)
+# @author Simon Heybrock
+from __future__ import annotations
+
+import functools
+import operator
+from collections.abc import MutableMapping
+from typing import Callable, Iterable
+
+from .cpp_classes import DataArray, Dataset, Variable
+
+
+def _item_dims(item):
+    return getattr(item, 'dims', ())
+
+
+def _summarize(item):
+    if isinstance(item, DataGroup):
+        return f'{type(item).__name__}({len(item)}, {item.sizes})\n'
+    if hasattr(item, 'sizes'):
+        return f'{type(item).__name__}({item.sizes})\n'
+    return str(item)
+
+
+class DataGroup(MutableMapping):
+    """
+    A group of data. Has dims and shape, but no coords.
+    """
+
+    def __init__(self, items=None):
+        self._items = {}
+        if items is not None:
+            for name, item in items.items():
+                self[name] = item
+
+    def __len__(self) -> int:
+        return len(self._items)
+
+    def __iter__(self):
+        yield from self._items
+
+    def __getitem__(self, name):
+        if isinstance(name, str):
+            return self._items[name]
+        if name == ():
+            return DataGroup({key: var[()] for key, var in self.items()})
+        dim, index = name
+        if isinstance(index, int) and self.sizes[dim] is None:
+            raise ValueError(
+                f"Positional indexing dim {dim} not possible as the length is not "
+                "unique.")
+        return DataGroup({
+            key: var[dim, index] if dim in _item_dims(var) else var
+            for key, var in self.items()
+        })
+
+    def __setitem__(self, name, value):
+        self._items[name] = value
+
+    def __delitem__(self, name: str):
+        del self._items[name]
+
+    @property
+    def dims(self):
+        dims = ()
+        for var in self.values():
+            # Preserve insertion order
```

Comment:
The order can be strange. Say item 1 has dims (x, y) and item 2 has (z, y, x), the result is (x, y, z). I think this is fine as long as all properties use the same order and it is clear that it need not correspond to any item (except the first whose dims are always in order).

What I want to say with this is, why bother with preserving the order beyond making sure that `dims` and `shape` return the same?

Reply:
I think mainly to make it predictable. Might not be too important in practice. I think in particular when `DataGroup` would usually be used with items with the same dims this could be useful, e.g., because it would then ensure that dims of the data group match dims of the items.

Reply:
So you would promise to always produce a certain order?

Reply:
Well... Hyrums law, i.e., unless we do, things will nevertheless depend on the implementation detail. So we have to decide which order we want. Another option could be alphabetical, but I felt that would be confusing, in particular for data groups with a single item, or only items with matching dims (since sorting could reverse the dim order).